breakpad-symbols/src/http.rs

```diff
@@ -142,46 +151,70 @@ fn file_key(module: &(dyn Module + Sync), file_kind: FileKind) -> FileKey {
     (module_key(module), file_kind)
 }
 
-fn create_cache_file(tmp_path: &Path, final_path: &Path) -> io::Result<NamedTempFile> {
-    // Use tempfile to save things to our cache to ensure proper
-    // atomicity of writes. We may want multiple instances of rust-minidump
-    // to be sharing a cache, and we don't want one instance to see another
-    // instance's partially written results.
-    //
-    // tempfile is designed explicitly for this purpose, and will handle all
-    // the platform-specific details and do its best to cleanup if things
-    // crash.
-
+async fn create_cache_file(final_path: &Path) -> io::Result<PartialDownload> {
     // First ensure that the target directory in the cache exists
     let base = final_path.parent().ok_or_else(|| {
         io::Error::new(
             io::ErrorKind::Other,
             format!("Bad cache path: {:?}", final_path),
         )
     })?;
-    fs::create_dir_all(&base)?;
-
-    NamedTempFile::new_in(tmp_path)
+    fs::create_dir_all(&base).await?;
+
+    // Try to create a new file in the final destination with `.part` added to the extension.
+    // Use an attempt counter to vaguely allow multiple processes to race on this, without them
+    // having to cooperate in any way. If there's only one instance this will just work reasonably.
```

Comment:
So historically we've used this pattern of creating a tempfile somewhere else and swapping it in. This is problematic for a few reasons:

* The tempfile crate is not integrated with tokio::async
* Moving files between directories has the potential to run into "different file system" problems
* There is no signal between separate processes that a file is already being downloaded, so they will race to download and swap in the same file

This new approach is more similar to how e.g. Firefox downloads a file, in that it's created where it wants to go with a special suffix and then renamed when done. Now *ideally* separate processes could use this to detect an ongoing download and block/collaborate on this, but that has some complexities I am not in the mood to sort out right now. In the ideal version of this design we move to an architecture where you just spawn one instance on a server and it knows how to use all the system's parallelism!

(I think you could maybe use os file locking in some clever way, idk)

So for now if a .part is detected, this implementation just repeatedly tries different names until it gives up. This isn't really shippable.